src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+        for (GATKRead r : shard.reads) {
+            SimpleInterval readInterval = new SimpleInterval(r);
+            ArrayList<Variant> variantsOverlappingThisRead = shard.variantsOverlapping(readInterval);
+            // we pass all the bases. That's better because this way it's just a shared
+            // pointer instead of being an array copy. Downstream processing is fine with having
+            // extra bases (it expects a few, actually).
+            readContext.add(new ReadContextData(refBases, variantsOverlappingThisRead));
+        }
+        return shard.withReadContext(readContext);
+    }
+
+    protected static final class SubdivideAndFillReadsIterator implements Iterator<ContextShard>, Serializable {
+        private static final long serialVersionUID = 1L;
+        private static final Logger log = LogManager.getLogger(SubdivideAndFillReadsIterator.class);
+        private final transient SamReader reader;
+        private final transient SAMRecordIterator query;
+        private final ContextShard shard;
+        private final String bam;
+        private final ReadFilter optFilter;
+        private final int lastValidPos;
+        private final int firstValidPos;
+        private final List<SimpleInterval> subshards;
+        private int currentSubShardIndex;
+        private ArrayList<GATKRead> readsSoFar = new ArrayList<>();
+        private Storage.Objects storageClient = null;
+        private SimpleInterval currentSubShard;
+        private ContextShard nextOutput = null;
+        private boolean readerClosed = false;
+        private int outputCount;
+
+        public SubdivideAndFillReadsIterator(String bam, AuthHolder auth, int outputShardSize, int margin, final ReadFilter optFilter, ContextShard shard) throws IOException, GeneralSecurityException, ClassNotFoundException {
+            this.bam = bam;
+            this.shard = shard;
+            this.optFilter = optFilter;
+            lastValidPos = shard.interval.getEnd() + margin;
+            firstValidPos = shard.interval.getStart() - margin;
+            ArrayList<SimpleInterval> ints =new ArrayList<>();
+            ints.add(shard.interval);
+            subshards = IntervalUtils.cutToShards(ints, outputShardSize);
+            currentSubShardIndex = 0;
+            currentSubShard = subshards.get(currentSubShardIndex);
+
+            if (BucketUtils.isCloudStorageUrl(bam)) {
+                storageClient = auth.makeStorageClient();
+                reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+            } else if (BucketUtils.isHadoopUrl(bam)) {
+                throw new RuntimeException("Sorry, Hadoop paths aren't yet supported");
+            } else {
+                // read from local file (this only makes sense if every worker sees the same thing, e.g. if we're running locally)
+                reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
```

Comment:
This also won't work with cram files (as they require a reference to be passed in to the `SamReaderFactory`) -- could you mention this in the class docs?